### PR TITLE
Black magic probe issue #183

### PR DESCRIPTION
**Pull request by micooke**
In support of 'issue' #183 - Add Black Magic Probe (BMP) upload method  …
This is basically a merge of work done by @rogerclarkmelbourne in arduino-nRF5-customised, with the addition of bootloader uploading as per @sandeepmistry openocd reference in platform.txt
Note: This incorporates changes based off feedback from @sandeepmistry, @dmikhalsky, @dlabun
* README.md - BMP driver installation details
* platform.txt - BMP tool setup
* programmers.txt - Add BMP as a programmer option

### CODE REVIEW

**Comment by micooke** [comment]
Tidy-up of #186

File: programmers.txt [hunk 1]
```diff
@@ -18,3 +18,11 @@ cmsisdap.protocol=cmsis-dap
 cmsisdap.program.protocol=cmsis-dap
 cmsisdap.program.tool=openocd
 cmsisdap.program.setup_command=;
+
+bmp.name=Black Magic Probe (GDB)
```
Comment (sandeepmistry) [inline programmers.txt]: Can we change `bmp` to `blackmagicprobe` in this file?

File: platform.txt [hunk 2]
```diff
@@ -129,3 +129,27 @@ tools.openocd.erase.pattern=
 tools.openocd.bootloader.params.verbose=-d2
 tools.openocd.bootloader.params.quiet=-d0
 tools.openocd.bootloader.pattern="{path}/{cmd}" {bootloader.verbose} -f interface/{program.protocol}.cfg -c "{program.setup_command}" -f target/{upload.target}.cfg -c "init; halt; nrf51 mass_erase; program {{{runtime.platform.path}/cores/nRF5/SDK/components/softdevice/{softdevice}/hex/{softdevice}_{upload.target}_{softdeviceversion}_softdevice.hex}} verify reset; shutdown;"
+
+#
+# blackmagic probe upload
+#
+tools.blackmagicprobe.path={runtime.tools.gcc-arm-none-eabi-5_2-2015q4.path}/bin/
```
Comment (sandeepmistry) [inline platform.txt]: Can you replace the `{runtime.tools.gcc-arm-none-eabi-5_2-2015q4.path}/bin/` with `{compiler.path}` here?

That way if gcc is upgraded, only one line needs to be updated:
```
compiler.path={runtime.tools.gcc-arm-none-eabi-5_2-2015q4.path}/bin/
```

Comment (micooke) [reply platform.txt]: unfortunately it looks like {compiler.path} is not visible to tools, only to the compiler itself and recipes.

File: programmers.txt (hunk 1 above)
Comment (micooke) [reply programmers.txt]: Fixed

File: platform.txt (hunk 2 above)
Comment (sandeepmistry) [reply platform.txt]: Confirmed.

**Comment by sandeepmistry** [comment]
I'm going to merge this for now.

If I try to upload on a non-existent serial port on my Mac, the tool reports an error, but IDE reports `Done uploading.`. @micooke it would be nice to fix this ...

```
/Users/smistry/Library/Arduino15/packages/sandeepmistry/tools/gcc-arm-none-eabi/5_2-2015q4/bin/arm-none-eabi-gdb -quiet -cd /var/folders/5d/pmvtw7p579n5npm6p421qf540000gn/T/arduino_build_708668 -b 230400 -l 10 -ex set debug remote 0 -ex set target-async off -ex set remotetimeout 10 -ex set mem inaccessible-by-default off -ex set confirm off -ex set height 0 -ex target extended-remote \\.\/dev/cu.Bluetooth-Incoming-Port -ex monitor swdp_scan -ex attach 1 -ex file Blink.ino.elf -ex load -ex tbreak main -ex run -ex echo \nProgram complete!\n -ex quit 
\\.\/dev/cu.Bluetooth-Incoming-Port: No such file or directory.
"monitor" command not supported by this target.
Don't know how to attach.  Try "help target".
You can't do that when your target is `exec'
Don't know how to run.  Try "help target".
Reading symbols from Blink.ino.elf...done.
Temporary breakpoint 1 at 0x422: file /Users/smistry/Documents/Arduino/hardware/sandeepmistry/nRF5/cores/nRF5/main.cpp, line 29.

Program complete!
```

**Comment by micooke** [comment]
@sandeepmistry - sorry i didnt see this comment until @probonopd pointed this out.
Unfortunately i couldn't figure out how to conditionally write a Program success/failed (that was portable) - so i stuck with a 'Program complete' message.

**Pull request merged**